### DISCUSSION TITLE
API Mocks

### DISCUSSION THREAD

**Discussion by jachin** [modernweb-dev/web#1437] [Q&A]
hi there

I'm setting up `@web/dev-server` for a small-ish project that used webpack before. I'm trying to figure out the best way to hook up some mock API responses.

I've tried using a middleware proxy but, I'm not sure exactly what I want. It looks very straight forward to send (some) requests (based on the pattern of the URL) to a different server. However what I want to do _seems_ like it's simpler, which is just server up json files.

I've tried [rewriting the requests](https://modern-web.dev/docs/dev-server/middleware/#rewriting-request-urls), but I'm getting errors about changing headers after the body had already been sent.

I _think_ I could figure out _a_ solution to this using some [KOA middleware](https://github.com/koajs/koa/wiki#middleware), but it all seem more complicated than it should be.

Does anyone have an example handy they could point me to?

Thanks

**Comment by codydubat** [comment]
Hello! Do you have any update or have you found any solution on this? I have some middleware that serves json files well for web-dev-server, but I can't figure out why it doesn't work for the test runner. I get 404 when I try to call the same mock APIs from web-test-runner.

**Reply by jachin** [reply]
I'm afraid I have not. I went back and looked at what I think was the web component I was working on at the time, just to see, other people have worked on it in the mean time but I don't think we ever figured anything out.

It sounds like you got further than I did.
